proposals/0298-bank-hash-in-block-footer.md

```diff
@@ -0,0 +1,107 @@
+---
+simd: '0298'
+title: Bank Hash in Block Footer
+authors:
+  - Max Resnick
+category: Standard
+type: Core
+status: Idea
+created: 2024-03-26
+---
+
+## Summary
+
+This proposal adds the bank hash to the block footer and introduces
+validity constraints to ensure the bank hash matches the expected value.
+
+## Motivation
+
+This proposal is part of the changes required for Alpenglow. In Alpenglow,
+votes do not contain the `bank_hash`. Instead they are purely on `block_id`
+If we remove `bank_hash` from votes without any other
+changes, then execution would not be a part of consensus at all, meaning if
+there is a bug in the runtime which causes a `bank_hash` mismatch it could be
+difficult to identify quickly and resolve. This proposal moves consensus on the
+`bank_hash` to the block footer so that we still have consensus
+on execution state.
+
+## New Terminology
+
+Block Footer: This is an umbrella term for information contained in the last
+FEC set.
+The exact structure is described in SIMD 0307.
+
+## Detailed Design
+
+### Block Footer Changes
+
+The block footer structure outlined in SIMD 0307
+will be extended to include a new field:
+
+```rust
+pub struct BlockFooter {
+    // ... existing fields ...
+    pub bank_hash: Hash,  // Hash of the block's bank state
+}
+```
+
+### Validity Constraints
+
+In addition to the current validity constraints, a block will be considered
+invalid if:
+
+1. The `bank_hash` does not match the post execution state of the block
+2. The `bank_hash` is missing (for blocks after the feature is activated)
+
+### Implementation Details
+
+1. The `bank_hash` will be computed at the end of packing each block.
+The validator proposing the block will include it in the block footer.
+2. Replaying validators will verify the `bank_hash` in the block footer
+matches their computed hash of the state after executing the block.
+This is similar to what they do with the `bank_hash`
+contained in votes today.
```

Comment:
can you specify what validators should do if the bank hash doesn't match? eg. should they vote "skip" in alpenglow?

Reply:
Yeah I think explicitly we should call out that the bank will be marked as dead and alpenglow validators should vote skip for this block (and any descendants)